### PR TITLE
Feature/lndp noahmp2 : add land perturbation scheme for Noah-MP fractional veg

### LINKED ISSUE #1076: Update land perturbation scheme to work with Noah-MP
## Description
The current land perturbation scheme (lndp=2) needs to be updated to the Noah-MP land model for use in GFSv17. 

## Solution
Update the routine applying land perturbations to also work with Noah-MP. 
At the same time, revise the namelist options for applying land perturbations to the regional system (HRRR), since these are currently confusing and sometimes use the land model as a proxy for a regional / global run. 

## Alternatives
None.

## Related to

Issues: 
ufs-weather-model  #1076 
stochastic_physics #55 
fv3atm #493

### PR DESCRIPTION
**Pull request by ClaraDraper-NOAA**
# PR Checklist

- [ x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [ x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [ x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [ x] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.

## Description

Updates to allow land perturbation scheme to be applied for Noah-MP. Currently, for Noah-MP model parameters can perturb only vegetation fraction.

Main code changes:
-added lsm_noahmp model option to lndp_type==2 land perturbation scheme (fv3-atm, stochastic_physics)
-updated comments
-cleaned up the namelists to make the different options clearer. Removed misleading lndp_each_step variable from gfs_physics_nml, and replaced it with "lndp_model_type" in the nam_sfcperts namelist for the different forecast types (cycling DA, short forecasts, perturbing only initial conditions) (fv3-atm, stochastic_physics, ufs-weather-model)
NOTE: see note below, re: specification of smc perturbation for RAP/HRRR/etc (lndp_model_type==2)
-revised code in stochastic_physics_wrapper to only allocate arrays that will be used when passed into lndp_apply_perts (fv3-atm)
-deleted unused albedo arrays in stochastic_physics_wrapper and lndp_apply_perts (fv3-atm, stochastic_physics)
-fixed bug causing precision errors when calculating the soil moisture ice content (often ~0.0; stochastic_physics)
-fixed bug in which lndp_apply_perts wasn't called for nscyc = 0. (fv3-atm)
-increased max_nvar_lndp at Jeff Ator's request (fv3-atm)
-added a new test for the Noah-MP and lndp==2 combination (ufs-weather-model)

### Issue(s) addressed

Addresses issue https://github.com/NOAA-EMC/fv3atm/pull/55 in stochastic_physics.

## Testing

All tests passed on here, except the lndp ones (see below).

Changes to tests:
-added new control_p8_lndp to test Noah-MP land perts
-updated previous lndp tests with new results, now that precision errors when calculating the soil moisture ice content is fixed
-also, for rap_lndp_debug the perturbation to smc at the initial conditions (applied only once) was previously specified as a rate, then divided by the time step before being applied. This makes the magnitude of perturbation dependent on the model time step. This has been fixed with the introduction of the lndp_model_type==2 option for regional sysyems, so that the perturbation is now specified as an absolute amount (changed 0.2 to 0.017 in the test). The previously used smc pert needs to be multiplied by timestep/3600 to get the same results with the new code.

New tests are here on hera:
/scratch2/BMC/gsienkf/Clara.Draper/stmp4/Clara.Draper/FV3_RT/REGRESSION_TEST_INTEL

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [x] opnReqTest for newly added/changed feature
- [x] CI https://github.com/ufs-community/ufs-weather-model/actions/runs/2226992322

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

Three linked PRS:
stochastic_physics ( Feature/lndp noahmp2 : add land perturbation scheme for Noah-MP fractional veg https://github.com/noaa-psd/stochastic_physics/pull/56)
ufs-weather-model (add land perturbation scheme for Noah-MP fractional veg #1143 )
fv3-atm ( Feature/lndp noahmp2 : add land perturbation scheme for Noah-MP fractional veg https://github.com/NOAA-EMC/fv3atm/pull/513)

### CODE REVIEW

**Comment by junwang-noaa** [comment]
@ClaraDraper-NOAA Is this PR required for P8c or any other applications?

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Is this PR required for P8c or any other applications?

@junwang-noaa Sorry for the late reply, I was on leave. This is intended to be a part of the stochastic physics scheme, but I'm not sure when that will go into the prototypes. @pjpegion ? Bing Fu is already using it for ensembles testing.

File: tests/tests/control_p8_lndp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                   RESTART/phy_data.tile3.nc \
+                   RESTART/phy_data.tile4.nc \
+                   RESTART/phy_data.tile5.nc \
+                   RESTART/phy_data.tile6.nc \
+                   RESTART/sfc_data.tile1.nc \
+                   RESTART/sfc_data.tile2.nc \
+                   RESTART/sfc_data.tile3.nc \
+                   RESTART/sfc_data.tile4.nc \
+                   RESTART/sfc_data.tile5.nc \
+                   RESTART/sfc_data.tile6.nc"
```
Comment (DusanJovic-NOAA) [inline tests/tests/control_p8_lndp]: There's no need to save all these RESTART/* files unless they are going to be used by some other restart test.

File: tests/parm/cpld_control_lndp.nml.IN [hunk 2]
```diff
@@ -0,0 +1,366 @@
+&atmos_model_nml
```
Comment (DusanJovic-NOAA) [inline tests/parm/cpld_control_lndp.nml.IN]: I wonder if this new namelist template file is needed or we can use (updated) parm/cpld_control.nml.IN? What's the difference between this template and cpld_control.nml.IN? Are lndp variables the only difference?

File: tests/tests/control_p8_lndp (hunk 1 above)
Comment (ClaraDraper-NOAA) [reply tests/tests/control_p8_lndp]: I've deleted the RESTART lines.

File: tests/parm/cpld_control_lndp.nml.IN (hunk 2 above)
Comment (ClaraDraper-NOAA) [reply tests/parm/cpld_control_lndp.nml.IN]: I generalised the cpld_control.nml.IN and deleted cpld_control_lndp.nml.IN.

File: tests/tests/control_p8_lndp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+export USE_CICE_ALB=.false.
+
+export WRITE_NSFLIP=.true.
+
+export FV3_RUN=control_run.IN
+export CCPP_SUITE=FV3_GFS_v17_p8
+export FIELD_TABLE=field_table_thompson_noaero_tke
+export DIAG_TABLE=diag_table_gfsv16_merra2
+# use same namelist for standalone,coupled P7
+export INPUT_NML=cpld_control_lndp.nml.IN
```
Comment (DusanJovic-NOAA) [inline tests/tests/control_p8_lndp]: I believe this must be updated now that cpld_control_lndp.nml.IN is removed. Please double check.

Comment (ClaraDraper-NOAA) [reply tests/tests/control_p8_lndp]: Done, thanks.

**Comment by junwang-noaa** [comment]
@ClaraDraper-NOAA Does the [control_p8_lndp](https://github.com/ufs-community/ufs-weather-model/pull/1143/files#diff-fccaecfc33bcf16be61dbad68cbffceeac3cd9d94548a2985118db2c99fcbadf) test reproduce with threading, decomposition, restart and does it run with debug mode? You can reference the settings in control_p8 (control_2threads_p8, control_decomp_p8, control_restart_p8, control_debug_p8) to run the test. Thanks.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Does the [control_p8_lndp](https://github.com/ufs-community/ufs-weather-model/pull/1143/files#diff-fccaecfc33bcf16be61dbad68cbffceeac3cd9d94548a2985118db2c99fcbadf) test reproduce with threading, decomposition, restart and does it run with debug mode? You can reference the settings in control_p8 (control_2threads_p8, control_decomp_p8, control_restart_p8, control_debug_p8) to run the test. Thanks.

@junwang-noaa To make sure I understand, you'd like me to tun the control_p8_lndp test, using each of threading, decomposition, restart, and debugging (looking at the tests you pointed me to to figure out how to run each one), and make sure that I get the same result as the standard test with each of these options? 

Clara

**Comment by junwang-noaa** [comment]
That is correct.

File: tests/tests/control_p8_lndp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+export NPZP=128
+export DT_ATMOS=720
+export SYEAR=2021
+export SMONTH=03
+export SDAY=22
+export SHOUR=06
+export RESTART_INTERVAL="12 -1"
+export OUTPUT_GRID='gaussian_grid'
+export WRITE_DOPOST=.true.
+export FHMAX=48
```
Comment (DusanJovic-NOAA) [inline tests/tests/control_p8_lndp:39]: Since only 0, 21 and 24h' history files are saved as baselines, there's no need to run 48h forecast.

Comment (ClaraDraper-NOAA) [reply tests/tests/control_p8_lndp:39]: I'm going to change this to add testing of the 48 hr output, as I want the test to check that the global_cycle update at 24 hours (FHCYC) is handled properly.

Comment (DusanJovic-NOAA) [reply tests/tests/control_p8_lndp:39]: Ok. Thanks.

**Review by junwang-noaa — APPROVED** [review]

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: hera
Compiler: gnu
Job: BL
[BL] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model
[BL] Error: Test control_stochy 003 failed in run_test failed 
[BL] Error: Test control_stochy_debug 030 failed in run_test failed 
Please make changes and add the following label back: hera-gnu-BL

**Comment by ClaraDraper-NOAA** [comment]
> _**Automated RT Failure Notification**_ Machine: hera Compiler: gnu Job: BL [BL] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model [BL] Error: Test control_stochy 003 failed in run_test failed [BL] Error: Test control_stochy_debug 030 failed in run_test failed Please make changes and add the following label back: hera-gnu-BL

Brian - let me know if you want me to try to address this (in which case, I'd need help running gnu on hera). I ran the default tests on hera (intel) and this test passed.

**Comment by DusanJovic-NOAA** [comment]
> > _**Automated RT Failure Notification**_ Machine: hera Compiler: gnu Job: BL [BL] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model [BL] Error: Test control_stochy 003 failed in run_test failed [BL] Error: Test control_stochy_debug 030 failed in run_test failed Please make changes and add the following label back: hera-gnu-BL
> 
> Brian - let me know if you want me to try to address this (in which case, I'd need help running gnu on hera). I ran the default tests on hera (intel) and this test passed.

This is the error:

```
  0: At line 167 of file /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model/stochastic_physics/compns_stochy.F90
  0: Fortran runtime error: Cannot match namelist object name xxx
  0: 
  0: Error termination. Backtrace:
  0: #0  0x2b9173feab47 in finalize_transfer
  0:    at /tmp/Role.Apps/spack-stage/spack-stage-gcc-9.2.0-wqdecm4rkyyhejagxwmnabt6lscgm45d/spack-src/libgfortran/io/transfer.c:4072
  0: #1  0x13f0b62 in __compns_stochy_mod_MOD_compns_stochy
  0:    at /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model/stochastic_physics/compns_stochy.F90:167
  0: #2  0x13d372c in __stochy_data_mod_MOD_init_stochdata
  0:    at /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model/stochastic_physics/stochy_data_mod.F90:75
  0: #3  0x13cf7e7 in __stochastic_physics_MOD_init_stochastic_physics
  0:    at /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/889739205/20220426131516/ufs-weather-model/stochastic_physics/stochastic_physics.F90:96
  0: #4  0x13b1ad1 in __stochastic_physics_wrapper_mod_MOD_stochastic_physics_wrapper
```

**Comment by DusanJovic-NOAA** [comment]
In `input.nml` I see:
```
&nam_sfcperts
  lndp_type      = 0
  lndp_model_type = 0
  LNDP_TAU=21600,
  LNDP_LSCALE=500000,
  ISEED_LNDP=2010,
  lndp_var_list = XXX
  lndp_prt_list = -999
/
```
Value for `lndp_var_list` must be a string. Try to change:
```
export LNDP_VAR_LIST='XXX'
```
to:
```
export LNDP_VAR_LIST="'XXX'"
```
in `default_vars.sh`

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: cheyenne
Compiler: gnu
Job: BL
[BL] Repo location: /glade/scratch/epicufsrt/autort/tests/auto/pr/889739205/20220426134512/ufs-weather-model
Please make changes and add the following label back: cheyenne-gnu-BL

**Comment by ClaraDraper-NOAA** [comment]
The ORT tests have passed (hera, intel) , and I've rerun hera intel with this morning's code push. All tests expected to pass against the April 21 baseline did pass. I manually checked the four tests with changes (the four tests with "lndp" in the name) against the April 25 baseline, and the output is unchanged. I think everything is ready.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Pull request merged**